Mathlib/Data/Set/Function.lean

```diff
@@ -624,6 +624,45 @@ lemma bijOn_id (s : Set α) : BijOn id s s := ⟨s.mapsTo_id, s.injOn_id, s.surj
 theorem BijOn.comp (hg : BijOn g t p) (hf : BijOn f s t) : BijOn (g ∘ f) s p :=
   BijOn.mk (hg.mapsTo.comp hf.mapsTo) (hg.injOn.comp hf.injOn hf.mapsTo) (hg.surjOn.comp hf.surjOn)
 
+/-- If `f : α → β` and `g : β → γ` and if `f` is injective on `s`, then `f ∘ g` is a bijection
+on `s` iff  `g` is a bijection on `f '' s`. -/
+theorem bijOn_comp_iff_bijOn_image {α β γ : Type*} {f : α → β} {g : β → γ} {s : Set α}
+    (hf : InjOn f s) {t : Set γ} : BijOn (g ∘ f) s t ↔ BijOn g (f '' s) t := by
+  constructor
+  · rintro ⟨h₁, h₂, h₃⟩
+    refine ⟨mapsTo_image_iff.mpr h₁, InjOn.image_of_comp h₂, fun x hx ↦ ?_⟩
+    obtain ⟨y, hy, rfl⟩ := h₃ hx
+    exact ⟨f y, ⟨y, hy, rfl⟩, rfl⟩
+  · rintro ⟨h₁, h₂, h₃⟩
+    exact ⟨mapsTo_image_iff.mp h₁, InjOn.comp h₂ hf <| mapsTo_image f s,
+      SurjOn.comp h₃ <| surjOn_image _ _⟩
+
+/--
+If we have a commutative square
+
+α --f--> β
+|        |
+p₁       p₂
+|        |
+\/       \/
+γ --g--> δ
+
+and `f` induces a bijection from `s : Set α` to `t : Set β` then `g`
+induces a bijection from the image of `s` to the image of `t`, as long as `g` is
+is injective on the image of `s`.
+-/
+theorem bijOn_image_image {α β γ δ : Type*} {f : α → β} {p₁ : α → γ} {p₂ : β → δ}
+    {g : γ → δ} (comm : ∀ a, p₂ (f a) = g (p₁ a)) {s : Set α} {t : Set β}
```

Comment:
Please use `Function.Commute` here.

Reply:
Do you mean `Function.comp`?

Reply:
`Function.Commute` doesn't seem relevant. I avoided `Function.comp` only because in the proof it was more convenient to have it in the form I wrote. I don't know what the canonical way to say "this square commutes" away from the `Algebra` setting (where we use `IsScalarTower`, not usable here).

Reply:
Sorry, I don't know how I missed the fact that all 4 functions are different.